### PR TITLE
Fix KT-51804

### PR DESCRIPTION
**Pull request by k163377**
The issues fixed are as described in the ticket.
https://youtrack.jetbrains.com/issue/KT-51804

### CODE REVIEW

File: core/reflection.jvm/src/kotlin/reflect/jvm/internal/calls/InlineClassAwareCaller.kt [hunk 1]
```diff
@@ -96,6 +93,10 @@ internal class InlineClassAwareCaller<out M : Member?>(
 
             descriptor.valueParameters.mapTo(kotlinParameterTypes, ValueParameterDescriptor::getType)
         }
+
+        // If the default argument is set, ((paramSize / Int.SIZE_BITS) + 1) masks and one marker are added to the end of the argument.
+        val extraArgumentsTail = (if (isDefault) (kotlinParameterTypes.size / Int.SIZE_BITS) + 2 else 0) +
```
Comment (udalov) [inline core/reflection.jvm/src/kotlin/reflect/jvm/internal/calls/InlineClassAwareCaller.kt:98]: The correct formula is `(kotlinParameterTypes.size + Int.SIZE_BITS - 1) / Int.SIZE_BITS` which means to round the division result up to the nearest integer:
https://github.com/JetBrains/kotlin/blob/d61e7db937499b0a26efacb8c10c5f92ec512aa2/core/reflection.jvm/src/kotlin/reflect/jvm/internal/KDeclarationContainerImpl.kt#L236

With the current changes, `callBy` should fail for functions that take 32 * N parameters, for integer N. Please add a function with exactly 32 parameters to the test case.

**Review by udalov — CHANGES_REQUESTED** [review]

File: core/reflection.jvm/src/kotlin/reflect/jvm/internal/calls/InlineClassAwareCaller.kt (hunk 1 above)
Comment (k163377) [reply core/reflection.jvm/src/kotlin/reflect/jvm/internal/calls/InlineClassAwareCaller.kt]: Thanks for the review.
Fixed and added tests.

[952ab0a](https://github.com/JetBrains/kotlin/pull/4779/commits/952ab0adb76e99e9cf2c8f68c39a830bd6f9d633)
[0ec01f6](https://github.com/JetBrains/kotlin/pull/4779/commits/0ec01f656a2b42cbcde55ed6d47a36dead86f46c)

By the way, would it be better to make this logic common?

Comment (udalov) [reply core/reflection.jvm/src/kotlin/reflect/jvm/internal/calls/InlineClassAwareCaller.kt]: Well, it's just an integer division with rounding up, I'm not sure that it warrants commonization. :)

File: core/reflection.jvm/src/kotlin/reflect/jvm/internal/calls/InlineClassAwareCaller.kt [hunk 2]
```diff
@@ -96,6 +93,11 @@ internal class InlineClassAwareCaller<out M : Member?>(
 
             descriptor.valueParameters.mapTo(kotlinParameterTypes, ValueParameterDescriptor::getType)
         }
+
+        // If the default argument is set,
+        // (kotlinParameterTypes.size + Integer.SIZE - 1) / Integer.SIZE masks and one marker are added to the end of the argument.
+        val extraArgumentsTail = (if (isDefault) ((kotlinParameterTypes.size + 31) / Integer.SIZE) + 1 else 0) +
```
Comment (udalov) [inline core/reflection.jvm/src/kotlin/reflect/jvm/internal/calls/InlineClassAwareCaller.kt:99]:
```suggestion
        val extraArgumentsTail = (if (isDefault) ((kotlinParameterTypes.size + Integer.SIZE) / Integer.SIZE) + 1 else 0) +
```
Minor, but let's either use 31/32 in both places, or `Integer.SIZE` in both places. :)

**Comment by udalov** [comment]
Thank you for the contribution!

**Pull request merged**